File: src/falcon_core/math/domains/knob_domain.py
"""A domain for a knob to fall in a range."""

from typing import TYPE_CHECKING

from .base_labelled_domain import BaseLabelledDomain
from .constants import INSTRUMENT_TYPES
from .dependancies import Jsonable, Knob, Units
from .domain import Domain

if TYPE_CHECKING:
    from .dependancies import Instrument
    from .typing import Connection, SymbolUnit


class KnobDomain(BaseLabelledDomain[Knob], Jsonable):
    """A Knob with physical bounds."""

    def __init__(
        self,
        default_name: str,
        bounds: tuple[float, float],
        pseudo_name: "Connection | None" = None,
        instrument_type: "Instrument" = INSTRUMENT_TYPES.DC_VOLTAGE_SOURCE,
        lesser_bound_contained: bool = True,
        greater_bound_contained: bool = True,
        units: "SymbolUnit" = Units.VOLT,
        description: str = "",
    ) -> None:
        """Initialize the KnobDomain object.

        Args:
            default_name (str): The default name of the knob.
            bounds (tuple[float, float]): The bounds of the knob.
            pseudo_name (Connection | None): The pseudo name of the knob.
            instrument_type (type): The type of instrument.
            lesser_bound_contained (bool): Whether the lesser bound is contained.
            greater_bound_contained (bool): Whether the greater bound is contained.
            units (str): The units of the knob.
            description (str): The description of the knob.
        """
        label = Knob(
            default_name=default_name,
            pseudo_name=pseudo_name,
            instrument_type=instrument_type,
            units=units,
            description=description,
        )
        super().__init__(
            label=label,
            bounds=bounds,
            lesser_bound_contained=lesser_bound_contained,
            greater_bound_contained=greater_bound_contained,
        )

    @classmethod
    def from_knob_and_domain(
        cls,
        knob: Knob,
        domain: Domain,
    ) -> "KnobDomain":
        """Create a KnobDomain from a Knob and a Domain.

        Args:
            knob (Knob): The knob to create a domain from.
            domain (Domain): The domain to create a knob from.

        Returns:
            KnobDomain: The knob domain.
        """
        return cls(
            default_name=knob.default_name,
            bounds=domain.bounds,
            pseudo_name=knob.pseudo_name,
            instrument_type=knob.instrument_type,
            lesser_bound_contained=domain.lesser_bound_contained,
            greater_bound_contained=domain.greater_bound_contained,
            units=knob.units,
            description=knob.description,
        )

    @classmethod
    def from_domain(
        cls,
        domain: Domain,
        default_name: str,
        pseudo_name: "Connection | None" = None,
        instrument_type: "Instrument" = INSTRUMENT_TYPES.DC_VOLTAGE_SOURCE,
        units: "SymbolUnit" = Units.VOLT,
        description: str = "",
    ) -> "KnobDomain":
        """Create a KnobDomain from a Domain.

        Args:
            domain (Domain): The domain to create a knob from.
            default_name (str): The default name of the knob.
            pseudo_name (str): The pseudo name of the knob.
            instrument_type (type): The type of instrument.
            units (str): The units of the knob.
            description (str): The description of the knob.

        Returns:
            KnobDomain: The knob domain.
        """
        return cls(
            default_name=default_name,
            bounds=domain.bounds,
            pseudo_name=pseudo_name,
            instrument_type=instrument_type,
            lesser_bound_contained=domain.lesser_bound_contained,
            greater_bound_contained=domain.greater_bound_contained,
            units=units,
            description=description,
        )

    @classmethod
    def from_knob(
        cls,
        knob: Knob,
        bounds: tuple[float, float],
        lesser_bound_contained: bool = True,
        greater_bound_contained: bool = True,
    ) -> "KnobDomain":
        """Create a KnobDomain from a Knob.

        Args:
            knob (Knob): The knob to create a domain from.
            bounds (tuple[float, float]): The bounds of the knob.
            lesser_bound_contained (bool): Whether the lesser bound is contained.
            greater_bound_contained (bool): Whether the greater bound is contained.

        Returns:
            KnobDomain: The knob domain.
        """
        return cls(
            default_name=knob.default_name,
            bounds=bounds,
            pseudo_name=knob._pseudo_name,
            instrument_type=knob.instrument_type,
            lesser_bound_contained=lesser_bound_contained,
            greater_bound_contained=greater_bound_contained,
            units=knob.units,
            description=knob.description,
        )

    @property
    def knob(self) -> Knob:
        """Return the knob."""
        return self._label

    @property
    def domain(self) -> Domain:
        """Return the domain."""
        return Domain(
            bounds=self.bounds,
            lesser_bound_contained=self.lesser_bound_contained,
            greater_bound_contained=self.greater_bound_contained,
        )

File: src/math/domains/KnobDomain.cpp
#include "falcon_core/math/domains/KnobDomain.hpp"

#include "falcon_core/instrument_interfaces/names/Knob.hpp"

namespace falcon_core::math::domains {
KnobDomain::KnobDomain() = default;
KnobDomain::KnobDomain(
    const std::string&                                  default_name,
    const std::pair<double, double>&                    bounds,
    const physics::device_structures::BaseConnectionSP& psuedo_name,
    const instrument_interfaces::Instrument&            instrument_type,
    bool                                                lesser_bound_contained,
    bool                                                greater_bound_contained,
    const physics::units::SymbolUnitSP&                 units,
    const std::string&                                  description)
    : BaseLabelledDomain<instrument_interfaces::names::Knob>(
          bounds,
          std::make_shared<instrument_interfaces::names::Knob>(
              default_name, psuedo_name, instrument_type, units, description),
          lesser_bound_contained,
          greater_bound_contained) {}

std::shared_ptr<KnobDomain> KnobDomain::from_knob_and_domain(
    const instrument_interfaces::names::KnobSP& knob, const DomainSP& domain) {
  return std::make_shared<KnobDomain>(knob->default_name(),
                                      domain->bounds(),
                                      knob->pseudo_name(),
                                      knob->instrument_type(),
                                      domain->lesser_bound_contained(),
                                      domain->greater_bound_contained(),
                                      knob->units(),
                                      knob->description());
}
std::shared_ptr<KnobDomain> KnobDomain::from_domain(
    const DomainSP&                                     domain,
    const std::string&                                  default_name,
    const physics::device_structures::BaseConnectionSP& pseudo_name,
    const instrument_interfaces::Instrument&            instrument_type,
    const physics::units::SymbolUnitSP&                 units,
    const std::string&                                  description) {
  return std::make_shared<KnobDomain>(default_name,
                                      domain->bounds(),
                                      pseudo_name,
                                      instrument_type,
                                      domain->lesser_bound_contained(),
                                      domain->greater_bound_contained(),
                                      units,
                                      description);
}
std::shared_ptr<KnobDomain> KnobDomain::from_knob(
    const instrument_interfaces::names::KnobSP& knob,
    const std::pair<double, double>&            bounds,
    const bool                                  lesser_bound_contained,
    const bool                                  greater_bound_contained) {
  return std::make_shared<KnobDomain>(knob->default_name(),
                                      bounds,
                                      knob->pseudo_name(),
                                      knob->instrument_type(),
                                      lesser_bound_contained,
                                      greater_bound_contained,
                                      knob->units(),
                                      knob->description());
}
const std::shared_ptr<instrument_interfaces::names::Knob>& KnobDomain::knob()
    const {
  return this->label();
}
}  // namespace falcon_core::math::domains

CEREAL_REGISTER_TYPE(falcon_core::math::domains::KnobDomain)
CEREAL_REGISTER_POLYMORPHIC_RELATION(falcon_core::generic::Song,
                                     falcon_core::math::domains::KnobDomain)
